node/src/components/consensus/highway_core/state.rs

```diff
@@ -915,43 +924,41 @@ impl<C: Context> State<C> {
 
     /// Returns whether the unit with `hash0` sees the one with `hash1` (i.e. `hash0 ≥ hash1`),
     /// and sees `hash1`'s creator as correct.
-    fn _sees_correct(&self, hash0: &C::Hash, hash1: &C::Hash) -> bool {
-        hash0 == hash1 || self.unit(hash0).panorama._sees_correct(self, hash1)
+    pub(crate) fn sees_correct(&self, hash0: &C::Hash, hash1: &C::Hash) -> bool {
+        hash0 == hash1 || self.unit(hash0).panorama.sees_correct(self, hash1)
     }
 
     /// Returns whether the unit with `hash0` sees the one with `hash1` (i.e. `hash0 ≥ hash1`).
-    fn _sees(&self, hash0: &C::Hash, hash1: &C::Hash) -> bool {
-        hash0 == hash1 || self.unit(hash0).panorama._sees(self, hash1)
+    pub(crate) fn sees(&self, hash0: &C::Hash, hash1: &C::Hash) -> bool {
+        hash0 == hash1 || self.unit(hash0).panorama.sees(self, hash1)
     }
 
     // Returns whether the units with `hash0` and `hash1` see each other or are equal.
-    fn _is_compatible(&self, hash0: &C::Hash, hash1: &C::Hash) -> bool {
-        hash0 == hash1
-            || self.unit(hash0).panorama._sees(self, hash1)
-            || self.unit(hash1).panorama._sees(self, hash0)
+    fn is_compatible(&self, hash0: &C::Hash, hash1: &C::Hash) -> bool {
+        hash0 == hash1 || self.unit(hash0).panorama.sees(self, hash1)
     }
 
     /// Returns the panorama of the confirmation for the leader unit `vhash`.
+    ///
+    /// Returns `None` if we are not supposed to confirm the proposal.
     pub(crate) fn confirmation_panorama(
         &self,
         own_idx: ValidatorIndex,
-        vhash: &C::Hash,
-    ) -> Panorama<C> {
-        let unit = self.unit(vhash);
-        let mut panorama;
-        // TODO(HWY-167): Confirmation panorama.
-        if let Some(prev_hash) = self.panorama().get(own_idx).correct().cloned() {
-            let own_unit = self.unit(&prev_hash);
-            panorama = unit.panorama.merge(self, &own_unit.panorama);
-            panorama[own_idx] = Observation::Correct(prev_hash);
-        } else {
-            panorama = unit.panorama.clone();
+        uhash: &C::Hash,
+    ) -> Option<Panorama<C>> {
+        let proposal = self.unit(uhash);
+        if let Some(prev_hash) = self.citable_panorama().get(own_idx).correct() {
+            // If proposal doesn't see our own previous vote, don't confirm it.
+            if !proposal.panorama.sees(self, prev_hash) {
+                return None;
+            }
         }
-        panorama[unit.creator] = Observation::Correct(*vhash);
+        let mut panorama = proposal.panorama.clone();
+        panorama[proposal.creator] = Observation::Correct(*uhash);
```

Comment:
Ah, there's another duplication of `inclusive_panorama`! :slightly_smiling_face:

Reply:
Done in [`361bdd6` (#572)](https://github.com/CasperLabs/casper-node/pull/572/commits/361bdd69ea987f47669f8a3f482a76c986d78b63)